source/MRMesh/MRMeshSave.cpp

```diff
@@ -221,6 +227,61 @@ VoidOrErrStr toBinaryStl( const Mesh & mesh, std::ostream & out, ProgressCallbac
     return {};
 }
 
+VoidOrErrStr toAsciiStl( const Mesh& mesh, const std::filesystem::path& file, ProgressCallback callback )
+{
+    std::ofstream out( file, std::ofstream::binary );
+    if ( !out )
+        return unexpected( std::string( "Cannot open file for writing " ) + utf8string( file ) );
+
+    return toAsciiStl( mesh, out, callback );
+}
+
+VoidOrErrStr toAsciiStl( const Mesh& mesh, std::ostream& out, ProgressCallback callback )
+{
+    MR_TIMER;
+#define SOLID_NAME "mesh"
+#define ENDL "\r\n"
+
+    out << "solid " SOLID_NAME ENDL;
+    auto notDegenTris = getNotDegenTris( mesh );
+    const float trisNum = float( notDegenTris.count() );
+    int trisIndex = 0;
+    for ( auto f : notDegenTris )
+    {
+        VertId a, b, c;
+        mesh.topology.getTriVerts( f, a, b, c );
+        assert( a.valid() && b.valid() && c.valid() );
+        const Vector3f& ap = mesh.points[a];
+        const Vector3f& bp = mesh.points[b];
+        const Vector3f& cp = mesh.points[c];
+        Vector3f normal = cross( bp - ap, cp - ap ).normalized();
+        std::string s = fmt::format( "{:.6f} {:.6f} {:.6f}", normal.x, normal.y, normal.z );
+        out << "facet normal " << s << ENDL;
+        out << "outer loop" ENDL;
+        for ( const Vector3f& p : { ap, bp, cp } )
+        {
+            s = fmt::format( "{:.8f} {:.8f} {:.8f}", p.x, p.y, p.z );
```

Comment:
Why `{:8f}` and not simply `{}`?

Reply:
The sample file has 6 fractional digits in normals and 8 fractional digits in points. Should I conform to it?
`
solid mystl_ascii
facet normal -0.503692 0.122505 -0.855153
outer loop
vertex 4.99090000 -5.70400000 6.78470000
vertex 4.93090000 -5.69800000 6.82090000
vertex 4.93080000 -5.63070000 6.83060000
endloop
endfacet
facet normal -0.896109 -0.064620 0.439104
outer loop
vertex 4.93080000 -5.63070000 6.83060000
vertex 4.93090000 -5.69800000 6.82090000
vertex 6.91970000 -8.29500000 10.49740000
endloop
...
endfacet
endsolid mystl_ascii
`

Reply:
I think not, we should not write trailing zeros. Simple `{}` should be fine